pkg/personname/info.go

```diff
@@ -0,0 +1,150 @@
+//
+// - Alphabetic: How a name is formally spelled using a Phonetic alphabet.
+// - Ideographic: How a name is represented using ideograms / ideographs.
+// - Phonetic: How a name is pronounced.
+//
+// Each of these groups can be inspected to access their individual segments (family
+// name, Given name, etc.)
+type Info struct {
+	// Alphabetic group information about the Alphabetic group.
+	Alphabetic GroupInfo
+	// Ideographic group information about the Ideographic group.
+	Ideographic GroupInfo
+	// Phonetic group information about the Phonetic group.
+	Phonetic GroupInfo
+
+	// NoNullSeparators will remove repeated separators around null groups when
+	// calling String() if set to true.
+	NoNullSeparators bool
+}
+
+// WithNullSeparators returns a new Info object that will keep trailing separators
+// that surround both null groups AND group segments: (ex: 'Potter^Harry^^^==').
+func (info Info) WithNullSeparators() Info {
+	info.NoNullSeparators = false
+	info.Alphabetic.NoNullSeparators = false
+	info.Ideographic.NoNullSeparators = false
+	info.Phonetic.NoNullSeparators = false
+
+	return info
+}
+
+// WithoutNullSeparators returns a new Info object that will remove trailing
+// separators that surround both null groups AND group segments:
+// (ex: 'Potter^Harry').
+func (info Info) WithoutNullSeparators() Info {
+	info.NoNullSeparators = true
+	info.Alphabetic.NoNullSeparators = true
+	info.Ideographic.NoNullSeparators = true
+	info.Phonetic.NoNullSeparators = true
+
+	return info
+}
+
+// String returns the original Raw representation of the PN value, in
+// '[Alphabetic]=[Ideographic]=[Phonetic]' format.
+func (info Info) String() string {
+	groupStrings := []string{
+		info.Alphabetic.String(), info.Ideographic.String(), info.Phonetic.String(),
+	}
+
+	// If we are removing trailing emtpy, we are going to check if the last group is
+	// empty, then trim it off the end if it is, until we hit a non-empty slice.
+	if info.NoNullSeparators {
+		out := 3
+		for i := 2; i >= 0; i-- {
+			if groupStrings[i] == "" {
+				out = i
+			} else {
+				break
+			}
+		}
+
+		// Trim the groups by getting a slice of our slice.
+		groupStrings = groupStrings[0:out]
+	}
+
+	return strings.Join(groupStrings, groupSep)
+}
+
+// IsEmpty returns whether the PN value contains any actual information. This method
+// ignores separator, so both '' and '^^^^=^^^^=^^^^' would return true.
+func (info Info) IsEmpty() bool {
+	return info.Alphabetic.IsEmpty() &&
+		info.Ideographic.IsEmpty() &&
+		info.Phonetic.IsEmpty()
+}
+
+// Parse PN dicom value into informational value.
+func Parse(valueString string) (Info, error) {
+	groups := strings.Split(valueString, groupSep)
+
+	if len(groups) > 3 {
+		return Info{}, newErrTooManyGroups(len(groups))
+	}
+
+	info := Info{}
+
+	// Range over the groups and assign them based on index.
+	for i, groupString := range groups {
+		switch i {
+		case 0:
+			groupInfo, err := groupFromValueString(groupString, "Alphabetic")
```

Comment:
I'm not sure about passing around raw string literals like "Alphabetic" -- might be better for these to be enums or consts. btw, it doesn't seem like these are used for anything other than annotating error strings inside groupFromValueString? If that's true, then probably what we should do is remove that parameter, and wrap the error with the "Alphabetic" context here on L113, etc.